src/beam/cli/logs.py

```diff
@@ -0,0 +1,121 @@
+import click
+from beta9 import terminal
+from beta9.config import DEFAULT_CONTEXT_NAME, get_settings, load_config
+from websockets.sync.client import connect
+
+
+def get_setting_callback(ctx: click.Context, param: click.Parameter, value: Any):
+    return getattr(get_settings(), param.name) if not value else value
+
+
+@click.group()
+def common(**_):
+    pass
+
+
+@common.command(
+    name="logs",
+    help="Follow logs of a deployment or a task.",
+)
+@click.option(
+    "--task-id",
+    type=click.STRING,
+    required=False,
+    help="",
+)
+@click.option(
+    "--deployment-id",
+    type=click.STRING,
+    required=False,
+    help="",
+)
+@click.option(
+    "--lines",
+    "-n",
+    type=click.INT,
+    required=True,
+    default=10,
+    help="Number of lines back to start.",
+)
+@click.option(
+    "--host",
+    "realtime_host",
+    type=click.STRING,
+    required=False,
+    callback=get_setting_callback,
+    hidden=True,
+)
+@click.option(
+    "--config-path",
+    type=click.Path(),
+    required=False,
+    callback=get_setting_callback,
+    hidden=True,
+)
+def logs(
+    task_id: Optional[str],
+    deployment_id: Optional[str],
+    lines: int,
+    realtime_host: str,
+    config_path: str,
+):
+    if bool(deployment_id) == bool(task_id):
+        raise click.BadArgumentUsage(
+            "Must supply either --deployment-id or --task-id, but not both."
+        )
+
+    contexts = load_config(config_path)
+    context = contexts[DEFAULT_CONTEXT_NAME]
+
+    websocket_params = {
+        "uri": realtime_host,
+        "additional_headers": {"X-BEAM-CLIENT": "CLI"},
+    }
+
+    now = datetime.datetime.now(datetime.timezone.utc)
+    logs_before = json.dumps(
+        {
+            "token": context.token,
+            "streamType": "LOGS_STREAM",
+            "action": "LOGS_QUERY",
+            "stream": False,
+            "objectType": "BETA9_TASK" if task_id else "BETA9_DEPLOYMENT",
+            "objectId": task_id or deployment_id,
+            "size": lines,
+            "endingTimestamp": now.isoformat(),
+        }
+    )
+
+    logs_current = json.dumps(
+        {
+            "token": context.token,
+            "streamType": "LOGS_STREAM",
+            "action": "LOGS_ADD_STREAM",
+            "stream": True,
+            "objectType": "BETA9_TASK" if task_id else "BETA9_DEPLOYMENT",
+            "objectId": task_id or deployment_id,
+            "startingTimestamp": now.isoformat(),
+        }
+    )
+
+    with connect(**websocket_params) as w, terminal.progress("Streaming..."):
+        w.send(logs_before)
+        print_message(w.recv())
+
+        w.send(logs_current)
+        try:
+            while True:
+                print_message(w.recv())
+        except KeyboardInterrupt:
+            terminal.print("\rGoodbye! 👋")
+
+
+def print_message(msg: Union[str, bytes]) -> None:
+    data = json.loads(msg)
+    hits = sorted(data["logs"]["hits"]["hits"], key=lambda k: k["_source"]["@timestamp"])
```

Comment:
is this always guaranteed to exist in the response?

Reply:
I believe so.

Elasticsearch returns this when there are no hits.

```json
{
  "took": 10,
  "timed_out": false,
  "_shards": {
    "total": 353,
    "successful": 353,
    "skipped": 297,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 0,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  }
}
```

We parse it in rt-api with these structs

```go
type PartialElasticSearchHits struct {
	Hits []struct {
		Source struct {
			Timestamp string `json:"@timestamp"`
		} `json:"_source"`
	} `json:"hits"`
}

type PartialElasticSearchResults struct {
	Hits PartialElasticSearchHits `json:"hits"`
}
```

Because the elasticsearch json results do include an empty list for `hits` when there aren't any hits, Go should unmarshal it into an empty list vs a nil.

To be safe though, I will wrap accessing these nested keys in a try/except and if an exception occurs, i'll just print the data vs sorting and printing the specific keys.